Add dreame adapter to latest

TA2k opened the pull request.

TA2k committed:
Add dreame to latest

TA2k commented:
RE-CHECK!

TA2k commented:
RE-CHECK!

mcm1957 commented:
reminder 4.1.2025

mcm1957 commented:
To verify the Object structure of this adapter during REVIEW please export the object structure of a working installation and attach the file i. You find a guide how to export the object struture here: https://github.com/ioBroker/ioBroker.repochecker/blob/master/OBJECTDUMP.md

Thanks

TA2k commented:
[dreame.0.json](https://github.com/user-attachments/files/18539890/dreame.0.json)

mcm1957 commented:
@TA2k 

First of all - THANK YOU for the time and effort you spend to maintain this adapter.

I would like to give some feedback based on my personal oppinion. @Apollon77 might have additional suggestions or even a different oppinion to one or the other statement. Please feel free to contact him if you cannot follow my suggestions or want to discuss some special aspects.

- [ ] Readme.md should contain a link to the manufacturer and/or device

  The README.md of any adapter related to device or m ultiple devices of a manufacturer should contain a link the manufacturers website and/or to a description of the device. This is to anable new users to easily check wether they associate the adapter with the correct device(s). (See https://github.com/ioBroker/ioBroker.repositories#requirements-for-adapter-to-get-added-to-the-latest-repository paragraph 4.)

- [ ] setObjectNotExists is not synchronous

  setObjectNotExists is an asynchronous function despite its name. Only the callback (not used by your code) is guaranteed to be executed after creation of the object. So please evalute wether ist really save to use setObjectNotExists without a callback. You may use setObjectNotExistsAsync with await or use await together with setObjectNotExists when reuiring js.controller 7.0.4 minimum (https://github.com/ioBroker/ioBroker.js-controller/blob/master/CHANGELOG.md#704-2024-12-04---lucy)

  Please adapt as most likely no other mechanismen than timings will gurantee that objects are really cerated befor first usage.

- [ ] setState is not synchronous
  
  setState is an asynchronous function too. The ste need not be store when the function returns. Please evaluate if an asynchronous operation is ok (setting an state fully asynchronous might be ok) or use await. see https://github.com/TA2k/ioBroker.dreame/blob/a7036514ce714d7ce40e0ded3337d382509818df/main.js#L825

- [ ] invalid characters should be filtered from object ids
  
  Some characters are not allowed to be used as part of an object id. If an object id is not hardcoded or even depending on user input you should ensure that such an object id does not contain an invalid character.  Iobroker provides the constant ```adapter.FORBIDDEN_CHARS```, i.e. by using the following snippet:

  ```
  function name2id(pName) {
      return (pName || '').replace(adapter.FORBIDDEN_CHARS, '_');
  }
  ```

  You retrieve the deviceId directly from the external cloud. This information is used as objectIds. So please filter the received information to ensure that no illegal characters ar contained.

- some objects are missing

  all intermediate objects must be created. 
   - `dreame.0.XXXXXXXX.general.deviceInfo.defaultVendors`
   - `dreame.0.XXXXXXXX.general.deviceInfo.extendScType` 
   - `dreame.0.XXXXXXXX.map.cleanset.2`
   - possibly more
   -  
  are not created. Please check / add creation.

- [ ] reevaluate state roles

  Only the values specified here (https://github.com/ioBroker/ioBroker.docs/blob/master/docs/en/dev/stateroles.md) may be used as state roles. Do not invent new names as this might disturb the functionalyity of other adapters or vis.

  In addition the roles MUST match the read/write functionality. As an example a role value.* requires that the state is a readable state. Input states (write only) should have roles like level.*. Please read the description carefully. States with role 'button' should (must) have attribute 'common.read=false' set. A button ( "Taster" in german only triggers when you press but else has no state), so it should also be not readable. This is also like HomeMatic does it. A switch has clear states true or false and so can be read.

  Some notes:

 - XXXXXXXX.map.cleanset.2.RoomOrder is writeable, so it should be level (value required write:false)
 - role 'array' does not exist. Please change. 

- [ ] linter setup

  Please use standard iobroker linter setup if acceptable. This is located at @iobroker/eslint-config. See migration guide at https://github.com/ioBroker/ioBroker.eslint-config/blob/main/MIGRATION.md.

  This is a suggestion only and not blocking. Decision is up to maintainer.
  
  
Thanks for reading and evaluating this suggestions.
McM1957

**Please add a comment when you have reviewed and fixed the suggestionsor at least commented the suggestions and you think the adapter is ready for a re-review!**

reminder 1.2.2025

TA2k commented:
1. I will not add official links to vendors or photos to prevent copyright infringement this adapter is not related to the vendor I'm not allowed to use the website or photos
2. I increase min jscontroller to 7.0.4
3. I don't need the result or wait for the setState to not block the code I don't use await
4.  Added replace function for the cleanest part
5. This objects are created via json2iob or for all cleanser part from getType() function
6. Fixed
7. I use my own linter config

mcm1957 commented:
@TA2k 
Thanks for the fast reaction.

Please see my comments and the 2 open issues:

1. I will not add official links to vendors or photos to prevent copyright infringement this adapter is not related to the vendor I'm not allowed to use the website or photos
    OK - accepted
    
2. I increase min jscontroller to 7.0.4
    see below
    
3. I don't need the result or wait for the setState to not block the code I don't use await
    OK - already indiceted at review that this is OK too
    
4. Added replace function for the cleanest part
    OK
    
5. This objects are created via json2iob or for all cleanser part from getType() function
    see below
    
6. Fixed
    OK
   
7. I use my own linter config
   Completly OK as already notes in review
   

The following issues need more / still attention:

- [ ] setObjectNotExists is not synchronous

  setObjectNotExists is an asynchronous function despite its name. Only the callback (not used by your code) is guaranteed to be executed after creation of the object. So please evalute wether ist really save to use setObjectNotExists without a callback. You may use setObjectNotExistsAsync with await or use await together with setObjectNotExists when reuiring js.controller 7.0.4 minimum (https://github.com/ioBroker/ioBroker.js-controller/blob/master/CHANGELOG.md#704-2024-12-04---lucy)

  Please adapt as most likely no other mechanismen than timings will gurantee that objects are really cerated befor first usage.

  Maybe I did not explain clearly. Even with js.controller 7.0.4 setObjectNotExists need an await. So please add await there to make objectCreation synchronous.
  
  Early js-controller versions did not return any useful value from setObjectNotExists but provided the callback as am entrypoint which guranteed that the object was created. New js-controller versions of setObjectNotExists return a promise if used without callback and are therfor identical to setObjectNotExistsAsync. This functionality has been added with js-controller 7.0.4 (see mentioned CHangelog of js.controller). setObjectNotExists is never synchronous by itself but requires the await if you want / must asure that the object is really created. (Of course you could collect all promises and wait at one final place too). 

- [ ] some objects are missing

  all intermediate objects must be created.

  - [ ] dreame.0.XXXXXXXX.general.deviceInfo.defaultVendors
  - [ ] dreame.0.XXXXXXXX.general.deviceInfo.extendScType
  - [ ] dreame.0.XXXXXXXX.map.cleanset.2
  - [ ] possibly more
 
  are not created. Please check / add creation.

  It's ok that hose object are created by json2iob. But there seems to be a bug at this code. At least within the dump provided the mentioned objects are not created. You can check you system and look for the type column of those (and other) objects. This column must not be empty and editing the the object should not result in "object does not exist".
  
  Maybe some ids received frome xternal source do contain a dot and the json2iob code does not eitehr replce the dot with ie. an _ or split the id and create the intermediate (folder type ?) object.

  Heres what I get when loading you json provide earlier:

  Type is empty:
![image](https://github.com/user-attachments/assets/08be6665-3fc9-470e-9d52-c3b6500ad0da)

  Edit returns:
![image](https://github.com/user-attachments/assets/f040269e-ab2a-4625-bb86-e9e02dc49d61)

  **Please check / verify and provide a new object json after corrects.**

TA2k commented:
fixed await for channel creation

Channel creation is fixed in json2iob
[dreame.0 (1).json](https://github.com/user-attachments/files/18565207/dreame.0.1.json)

mcm1957 commented:
Thanks for fixing

mcm1957 commented:
This adapter has been released to latest repository and should be visible within 24h maximum.

Please create a thread at https://forum.iobroker.net/category/91/tester titled like "Test Adapter " to collect some user feedback and provide a link to this topic when requesting addition to stable repository later.

Note: If an other testing topic already exists, it' OK to continue using this topic too.

If you do not have write access to TESTER Section of forum, please contact HOMORAN at forum using PN/CHAT
